crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -984,86 +1034,65 @@ impl CumulocityConverter {
     ) -> Result<Vec<MqttMessage>, ConversionError> {
         match &channel {
             Channel::EntityMetadata => {
-                if let Ok(register_message) = EntityRegistrationMessage::try_from(message) {
-                    return self
-                        .try_register_entity_with_pending_children(&register_message)
-                        .await;
-                }
-                Err(anyhow!(
-                    "Invalid entity registration message received on topic: {}",
-                    message.topic.name
-                )
-                .into())
+                // No conversion done here as entity data messages must be converted using pending_entities_from_incoming_message
```

Comment:
```suggestion
                // No conversion done here as entity data messages must be converted using try_register_source_entities
```
I'm not fully happy with this special case for registration messages. I was forced to do this because of the differences in the way we process a user published registration message vs auto-generated registration message, where the auto-registered entity registration message created by a `health` message is skipped to avoid duplicate status messages which was the cause of #2902 and fixed in #2901.

Reply:
In a follow up refactoring, I think it would be good to remove such inconsistencies between auto vs explicit registration. I also think that all the auto-registration logic should be encapsulated inside the entity store making the c8y mapper fully unaware of it.